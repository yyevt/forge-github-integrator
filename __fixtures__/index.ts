export * from "./pulls";
export * from "./repos";