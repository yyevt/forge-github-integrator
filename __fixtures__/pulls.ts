import { PullPresentationDto } from "../src/common/PresentationTypes";

export const OPEN_PULL: PullPresentationDto = {
    issueKey: "KAN-1",
    issueIsClosed: false,
    number: 42,
    title: "Add dark mode support",
    url: "https://github.com/org/repo/pull/42",
    state: "OPEN",
    isDraft: false,
    mergeable: "MERGEABLE",
    merged: false,
    repoName: "design-system",
    repoIsDisabled: false,
    repoIsArchived: false,
    authorLogin: "octocat",
};

export const CLOSED_MERGED_PULL: PullPresentationDto = {
    issueKey: "DEF-456",
    issueIsClosed: true,
    number: 17,
    title: "Fix typo in README",
    url: "https://github.com/org/repo/pull/17",
    state: "CLOSED",
    isDraft: false,
    mergeable: "UNKNOWN",
    merged: true,
    repoName: "design-system",
    repoIsDisabled: false,
    repoIsArchived: false,
    authorLogin: "hubot",
};