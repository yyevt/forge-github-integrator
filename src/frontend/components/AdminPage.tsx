import {
    ErrorMessage,
    Form,
    FormFooter,
    FormHeader,
    FormSection,
    HelperMessage,
    Label,
    LoadingButton,
    RequiredAsterisk,
    Textfield,
    useForm,
    ValidMessage
} from "@forge/react";
import React, {useState} from "react";
import {ForgeGateway} from "../services/ForgeGateway";
import {showToast} from "./Utils";

type FormInputs = {
    token: string;
}

export const AdminPage = () => {
    const {register, getFieldId, handleSubmit, formState: {errors, touchedFields}} = useForm<FormInputs>();
    const [isSubmitting, setSubmitting] = useState(false);

    const onSubmit = async (data: FormInputs): Promise<void> => {
        setSubmitting(true);
        try {
            await new ForgeGateway().saveToken(data.token);
            showToast("success", "Success!", "success", "Token was successfully saved.");
        } catch (e) {
            console.error(e);
            showToast("err", "Error", "error", (e instanceof Error ? e.message : `Unknown error occurred: ${e}`));
        } finally {
            setSubmitting(false);
        }
    };

    const isTokenFieldValid = !errors["token"];
    const isTokenFieldTouched = Boolean(touchedFields["token"]);

    return (
        <>
            <Form onSubmit={handleSubmit(onSubmit)}>
                <FormHeader title="Please enter GitHub token">
                    Required fields are marked with an asterisk <RequiredAsterisk/>
                </FormHeader>
                <FormSection>
                    <Label labelFor={getFieldId("token")}>GitHub token <RequiredAsterisk/></Label>
                    <Textfield placeholder="Paste token here" {...register("token", {
                        required: true,
                        pattern: /^(gh[pousr]_[a-zA-Z0-9]{36,255})$/
                    })} />
                    {isTokenFieldValid ? (
                        isTokenFieldTouched ? (
                            <ValidMessage>Valid token format</ValidMessage>
                        ) : (
                            <HelperMessage>You can use classic tokens or fine-grained ones.</HelperMessage>
                        )
                    ) : (
                        <ErrorMessage>Invalid token format</ErrorMessage>
                    )}
                </FormSection>
                <FormFooter>
                    <LoadingButton appearance="primary" type="submit" isLoading={isSubmitting}>
                        Save
                    </LoadingButton>
                </FormFooter>
            </Form>
        </>
    );
}