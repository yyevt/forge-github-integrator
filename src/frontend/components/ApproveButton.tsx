import React, {useCallback, useState} from "react";
import {LoadingButton} from "@forge/react";
import {showToast} from "./Utils";
import {ForgeGateway} from "../services/ForgeGateway";
import {PullPresentationDto} from "../../common/PresentationTypes";

export const ApproveButton = ({pullRequest}: { pullRequest: PullPresentationDto }) => {
    const [isApproving, setIsApproving] = useState(false);

    const approve = useCallback(async (): Promise<void> => {
        setIsApproving(true);
        try {
            await new ForgeGateway().approvePullRequest(pullRequest);
            showToast("success", "Success!", "success", "Pull.ts request was successfully approved.");
        } catch (e) {
            console.error(e);
            showToast("err", "Error", "error", (e instanceof Error ? e.message : `Unknown error occurred: ${e}`));
        } finally {
            setIsApproving(false);
        }
    }, [pullRequest]);

    return (
        <LoadingButton
            isLoading={isApproving}
            onClick={approve}
        >
            Approve
        </LoadingButton>
    );
};