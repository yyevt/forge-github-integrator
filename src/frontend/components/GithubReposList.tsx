import {Box, EmptyState, Stack} from "@forge/react";
import React from "react";
import {RepoMainInfo} from "./RepoMainInfo";
import {RepoPullsList} from "./RepoPullsList";
import {RepositoryPresentationDto} from "../../common/PresentationTypes";

export const GithubReposList = ({repos}: { repos: ReadonlyArray<RepositoryPresentationDto> }) => {
    return (
        <>
            {repos.length === 0 ? (
                <EmptyState header="Have no repositories"/>
            ) : (
                repos.map(repo => (
                    <Box padding="space.300" backgroundColor="color.background.discovery">
                        <Stack>
                            <RepoMainInfo repo={repo}/>
                            <RepoPullsList
                                pulls={repo.pullRequests}
                                pullsTotalCount={repo.pullRequestsTotalCount}
                                isMergeAllowed={repo.isMergeCommitAllowed}
                            />
                        </Stack>
                    </Box>
                ))
            )}
        </>
    );
}