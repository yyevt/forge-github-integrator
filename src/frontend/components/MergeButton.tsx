import React, {useCallback, useState} from "react";
import {LoadingButton} from "@forge/react";
import {showToast} from "./Utils";
import {ForgeGateway} from "../services/ForgeGateway";
import {PullPresentationDto} from "../../common/PresentationTypes";

export const MergeButton = ({pullRequest, isDisabled}: { pullRequest: PullPresentationDto, isDisabled: boolean }) => {
    const [isMerging, setIsMerging] = useState(false);

    const merge = useCallback(async (): Promise<void> => {
        setIsMerging(true);
        try {
            await new ForgeGateway().mergePullRequest(pullRequest);
            showToast("success", "Success!", "success", "Pull.ts request was successfully merged.");
        } catch (e) {
            console.error(e);
            showToast("err", "Error", "error", (e instanceof Error ? e.message : `Unknown error occurred: ${e}`));
        } finally {
            setIsMerging(false);
        }
    }, [pullRequest]);

    return (
        <LoadingButton
            isLoading={isMerging}
            isDisabled={isDisabled}
            onClick={merge}
        >
            Merge
        </LoadingButton>
    );
};