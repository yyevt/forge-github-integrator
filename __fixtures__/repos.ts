import {CLOSED_MERGED_PULL, OPEN_PULL} from "./pulls";
import {RepositoryPresentationDto} from "../src/common/PresentationTypes";

export const ACTIVE_REPO: RepositoryPresentationDto = {
    name: "design-system",
    url: "https://github.com/org/design-system",
    ownerLogin: "org",
    languages: ["TypeScript", "CSS"],
    isMergeCommitAllowed: true,
    isDisabled: false,
    isArchived: false,
    visibility: "public",
    createdAt: "2023-05-01T12:00:00Z",
    pullRequestsTotalCount: 2,
    pullRequests: [OPEN_PULL, CLOSED_MERGED_PULL],
};

export const ARCHIVED_REPO: RepositoryPresentationDto = {
    name: "legacy-app",
    url: "https://github.com/org/legacy-app",
    ownerLogin: "org",
    languages: ["JavaScript"],
    isMergeCommitAllowed: false,
    isDisabled: true,
    isArchived: true,
    visibility: "private",
    createdAt: "2020-01-15T08:30:00Z",
    pullRequestsTotalCount: 0,
    pullRequests: [],
};