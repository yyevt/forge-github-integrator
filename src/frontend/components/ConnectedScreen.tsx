import React, {useEffect, useState} from "react";
import {SectionMessage, Spinner, Stack, Text} from "@forge/react";
import {ForgeGateway} from "../services/ForgeGateway";
import {GithubReposList} from "./GithubReposList";
import {RepositoryPresentationDto} from "../../common/PresentationTypes";

export const ConnectedScreen = () => {
    const [isLoading, setIsLoading] = useState(false);
    const [repos, setRepos] = useState<ReadonlyArray<RepositoryPresentationDto>>([]);
    const [fetchReposError, setFetchReposError] = useState("");

    useEffect(() => {
        (async () => {
            setFetchReposError("");
            setIsLoading(true);
            try {
                const info = await new ForgeGateway().getGithubReposWithPulls();
                setRepos(info);
            } catch (e) {
                setFetchReposError(e instanceof Error ? e.message : `Unknown error occurred: ${e}`);
            } finally {
                setIsLoading(false);
            }
        })();
    }, []);

    if (isLoading) {
        return <Spinner/>;
    }
    return (
        <Stack space="space.200">
            <SectionMessage appearance="warning">
                <Text>Only limited amount of repositories and pull-requests are shown: max 100 repos, max 100 pulls</Text>
            </SectionMessage>
            {fetchReposError && (
                <SectionMessage appearance="error">
                    <Text>{fetchReposError}</Text>
                </SectionMessage>
            )}
            <GithubReposList repos={repos}/>
        </Stack>
    );
}