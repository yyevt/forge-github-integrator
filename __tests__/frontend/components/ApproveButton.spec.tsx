import React from "react";
import renderer from "react-test-renderer";
import {ApproveButton} from "../../../src/frontend/components/ApproveButton";
import {OPEN_PULL} from "../../../__fixtures__";

jest.mock("@forge/bridge", () => ({
    invoke: jest.fn()
}));

describe("ApproveButton", () => {

    it("renders correctly", () => {
        const reactRenderer = renderer.create(<ApproveButton pullRequest={OPEN_PULL}/>);

        expect(reactRenderer.toJSON()).toMatchSnapshot();
    });
});