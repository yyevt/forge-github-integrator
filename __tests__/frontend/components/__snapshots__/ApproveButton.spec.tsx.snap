// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ApproveButton renders correctly 1`] = `
<LoadingButton
  isLoading={false}
  onClick={[Function]}
>
  Approve
</LoadingButton>
`;
